### PR TITLE
Gcp state tracking

### LINKED ISSUE #849: googleclouddisk job fails with multiple workers on same GCE host.
When running multiple (dockerized) workers on the same GCE host the googleclouddisk job will fail with the error message "RESOURCE_IN_USE_BY_ANOTHER_RESOURCE".

The issue is that when a worker on a GCE host tries to mount the disk the first time it will succeed and expose the disk through the mapped /dev/ system into the first worker container. It fails when a second worker on the same GCE host tries to mount the disk as it's already mounted to that GCE host by the first worker.

The error is confusing, it's in use by the same resource, not by another resource.

### PR DESCRIPTION
**Pull request by wajihyassine**
This PR closes #849 by enabling resource state tracking for certain `Evidence` types, which atm only GoogleCloudDisk is needed.
* Added two new config options `RESOURCE_FILE`, which is the file used for resource state tracking and `RESOURCE_FILE_LOCK` which will lock methods so no two workers are trying to write to the state file at the same time.
* Two new variables for the `Evidence` class, `resource_tracked`, which is a bool that indicates whether a given Evidence type should be tracked in state file and `resource_id` which is a unique ID that will be used to track resources.
* Passing in `task_id` into `Evidence.preprocess` and `Evidence.postprocess` so can be tracked in state file.
* Added logic in `Evidence.preprocess` and `Evidence.postprocess` to update/remove `resource_id` and/or `task_id` if a given `Evidence` type needs to be tracked.
* Created a new module in `turbinia.processors.resource_manager.py` that handles managing (updating/removing) the resource state file.
* Implicitly locked `GoogleCloudDisk._preprocess` due to issue with likely race condition when two workers attempt to attach the same disk, there may be some lag with how long it takes for the disk to attach so a second worker can see that it is not attached, while the first worker is still trying to attach. 
* Added logic in `TaskResult.close` where if exception is raised during `postprocess` call, in the Exception block, it'll still attempt to remove the `task_id` from the resource state file for better tracking. 
* Updated  one of tests in `worker_test.py` due to us needing to pass `task_id` and Task was failing.

### CODE REVIEW

**Review by aarontp — COMMENTED** [review]
LG!  Just a couple small things.

File: turbinia/processors/resource_manager.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  fh.close()
+
+
+def PostProcessResourceState(resource_id, task_id):
+  """Removes the Evidence resource_id and/or task_id from the state file
+       as it is no longer needed.
+    
+    Args:
+      resource_id (str): The unique id representing the resource being tracked.
+      task_id (str): The id of a given Task.
```
[the remark this answers is not in the document]
Comment (wajihyassine) [reply turbinia/processors/resource_manager.py:80]: Done!

File: turbinia/processors/resource_manager.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    """
+  ValidateStateFile()
+  json_load = json.load(open(config.RESOURCE_FILE))
+  isDetachable = False
+
+  # Either remove task id or remove resource id if it is last Task remaining.
+  if resource_id in json_load.keys():
+    vals = json_load[resource_id]
```
[the remark this answers is not in the document]
Comment (wajihyassine) [reply turbinia/processors/resource_manager.py:88]: Done!

File: turbinia/processors/resource_manager.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    Args:
+      resource_id (str): The unique id representing the resource being tracked.
+      task_id (str): The id of a given Task.
+    """
+  ValidateStateFile()
+  json_load = json.load(open(config.RESOURCE_FILE))
+
+  # Append task_id to existing resource else add new resource id.
+  if resource_id in json_load.keys():
```
[the remark this answers is not in the document]
Comment (wajihyassine) [reply turbinia/processors/resource_manager.py:62]: Done!

File: turbinia/processors/resource_manager.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  """Removes the Evidence resource_id and/or task_id from the state file
+       as it is no longer needed.
+    
+    Args:
+      resource_id (str): The unique id representing the resource being tracked.
+      task_id (str): The id of a given Task.
+    """
+  ValidateStateFile()
+  json_load = json.load(open(config.RESOURCE_FILE))
+  isDetachable = False
```
[the remark this answers is not in the document]
Comment (wajihyassine) [reply turbinia/processors/resource_manager.py:84]: Done!

File: turbinia/processors/resource_manager.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      fh.write("{}")
+    fh.close()
+
+  # Ensure file can be loaded as json object
+  try:
+    with open(config.RESOURCE_FILE) as fh:
+      json_load = json.load(fh)
+    fh.close()
+  except ValueError as e:
+    raise TurbiniaException('Can not load json from file.')
```
[the remark this answers is not in the document]
Comment (wajihyassine) [reply turbinia/processors/resource_manager.py:47]: Done!

File: turbinia/evidence.py [hunk 6]
```diff
@@ -693,12 +722,15 @@ def _preprocess(self, _, required_states):
...
 
-    if EvidenceState.ATTACHED in required_states:
-      self.device_path, partition_paths = google_cloud.PreprocessAttachDisk(
-          self.disk_name)
-      self.partition_paths = partition_paths
-      self.local_path = self.device_path
-      self.state[EvidenceState.ATTACHED] = True
+    # Explicitly lock this method to prevent race condition with two workers
+    # attempting to attach disk at same time, given delay with attaching in GCP.
+    with filelock.FileLock(config.RESOURCE_FILE_LOCK):
```
[the remark this answers is not in the document]
Comment (wajihyassine) [reply turbinia/evidence.py:728]: Per chat, there is some logic in `AttachDisk` that checks to see if already mounted and if so returns paths

File: turbinia/config/turbinia_config_tmpl.py [hunk 7]
```diff
@@ -51,6 +51,13 @@
 # Path to a lock file used for the worker tasks.
 LOCK_FILE = '%s/turbinia-worker.lock' % OUTPUT_DIR
 
+# Path to a resource state file used for tracking shared Evidence types.
+RESOURCE_FILE = '%s/turbinia-state.json' % OUTPUT_DIR
```
Comment (hacktobeer) [inline turbinia/config/turbinia_config_tmpl.py:55]: This should be a shared path between workers running on the same machine. So OUTPUT_DIR should be eg mapped into a containers from the host, else each worker will have their own resource lock file. 
We should document this in one of the installation files and/or add a more descriptive comment here.

Comment (hacktobeer) [reply turbinia/config/turbinia_config_tmpl.py:55]: Same for RESOURCE_FILE_LOCK

**Review by hacktobeer — COMMENTED** [review]
minor comments, lgtm.

Can you also create a different PR to add the new config vars to below file?
https://github.com/forseti-security/osdfir-infrastructure/blob/master/modules/turbinia/templates/turbinia.conf.tpl

File: turbinia/config/turbinia_config_tmpl.py (hunk 7 above)
Comment (wajihyassine) [reply turbinia/config/turbinia_config_tmpl.py:59]: Thanks, updated install.md and config file to better explain resource state file!

**Review by hacktobeer — APPROVED** [review]

File: turbinia/processors/resource_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/processors/resource_manager.py]: Sorry, I wasn't very clear here.  What I meant was add `str(e)` to the message, like:
```
    raise TurbiniaException('Can not load json from file: {0!s}'.format(e))
```

File: turbinia/workers/__init__.py [hunk 8]
```diff
@@ -221,12 +222,15 @@ def close(self, task, success, status=None):
...
     if self.input_evidence:
       try:
-        self.input_evidence.postprocess()
+        self.input_evidence.postprocess(task_id=self.task_id)
       # Adding a broad exception here because we want to try post-processing
       # to clean things up even after other failures in the task, so this could
       # also fail.
       # pylint: disable=broad-except
       except Exception as exception:
+        with filelock.FileLock(config.RESOURCE_FILE_LOCK):
```
Comment (aarontp) [inline turbinia/workers/__init__.py:231]: Another super small thing, but maybe move this below the log line below just so when looking at the logs it's easier to understand why we are trying to remove the task from the resource a second time.

**Review by aarontp — COMMENTED** [review]
Just some small things for clearness/consistency.  Thanks!

**Comment by aarontp** [comment]
LGTM!

**Pull request merged**